src/main/java/ch/njol/skript/expressions/ExprBanData.java

```diff
@@ -0,0 +1,226 @@
+
+				"[the] reason %offlineplayer/string% was banned",
+				"[the] reason for %offlineplayer/string%'s ban",
+				"[the] ban reason of %offlineplayer/string%"
+		);
+	}
+
+	private BanEntryType entryType;
+	private Expression<Object> banTargetExpr; // offline player or string (ip)
+
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		entryType = BanEntryType.fromInt(matchedPattern);
+		banTargetExpr = (Expression<Object>) expressions[0];
+		return true;
+	}
+
+	@Override
+	protected Object @Nullable [] get(Event event) {
+		if (banTargetExpr == null)
+			return null;
+
+		Object banTarget = banTargetExpr.getSingle(event);
+		BanEntry<?> banEntry;
+
+		if (banTarget instanceof String ipTarget)
+			banEntry = getBanEntry(ipTarget);
+		else if (banTarget instanceof OfflinePlayer playerTarget)
+			banEntry = getBanEntry(playerTarget);
+		else
+			return null;
+
+		if (banEntry == null)
+			return null;
+
+		return switch (entryType) {
+			case BAN_DATE -> {
+				java.util.Date creation = banEntry.getCreated();
+				ch.njol.skript.util.Date skriptCreation = new ch.njol.skript.util.Date(creation.getTime());
+				yield new Object[]{ skriptCreation };
+			}
+			case SOURCE -> new Object[]{ banEntry.getSource() };
+			case EXPIRE_DATE -> {
+				java.util.Date expiration = banEntry.getExpiration();
+				if (expiration == null)
+					yield null;
+				ch.njol.skript.util.Date skriptExpiration = new ch.njol.skript.util.Date(expiration.getTime());
+				yield new Object[]{ skriptExpiration };
+			}
+			case REASON -> new Object[]{ banEntry.getReason() };
+		};
+	}
+
+	@Override
+	public boolean isSingle() {
+		return true;
+	}
+
+	@Override
+	public Class<?> getReturnType() {
+		return switch (entryType) {
+			case BAN_DATE, EXPIRE_DATE -> ch.njol.skript.util.Date.class;
+			case SOURCE, REASON -> String.class;
+		};
+	}
+
+	@Override
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		if (mode == ChangeMode.SET) {
+			return switch (entryType) {
+				case EXPIRE_DATE -> new Class<?>[]{ ch.njol.skript.util.Date.class };
+				case SOURCE, REASON -> new Class<?>[]{ String.class };
+				default -> null;
+			};
+		} else if (mode == ChangeMode.ADD || mode == ChangeMode.REMOVE) {
+			if (entryType == BanEntryType.EXPIRE_DATE)
+				return new Class<?>[]{ Timespan.class };
+		}
+		return null;
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		Object banTarget = banTargetExpr.getSingle(event);
+		BanEntry<?> banEntry;
+
+		if (banTarget instanceof String ipTarget)
+			banEntry = getBanEntry(ipTarget);
+		else if (banTarget instanceof OfflinePlayer playerTarget)
+			banEntry = getBanEntry(playerTarget);
+		else
+			return;
+
+		if (banEntry == null)
+			return;
+
+		if (mode == ChangeMode.SET) {
+			switch (entryType) {
+				case SOURCE -> {
+					String newSource = (String) delta[0];
+					if (newSource == null)
```

Comment:
delta won't contain null elements, so you can remove these (and assert instead if necessary)